api/api.proto

```diff
@@ -2694,3 +2698,72 @@ message LocateResponse {
     // <containerid>: /var/www
     map<string, string> dockerids = 2;
 }
+
+message VolumePlacementStrategy {
+  // Defines a list of rules as part of the placement spec. All the rules specified will
+  // be applied for volume placement.
+  // Rules that have enforcement as "required" are strictly enforced while "preferred" are best effort.
+  // In situations, where 2 or more rules conflict, the weight of the rules will dictate which wins.
+  repeated VolumePlacementRule rules = 1;
+}
+
+// This defines a single placement rule
+message VolumePlacementRule {
+  // Defines the number of volume replicas affected by this rule. If not provided,
+  // rule would affect all replicas
+  // (optional)
+  int32 affected_replicas = 1;
+  // Defines the weight of the rule which allows to break the tie with other matching rules. A rule with
+  // higher weight wins over a rule with lower weight.
+  // (optional)
+  int64 weight = 2;
+
+  // Defines the types of enforcement on the given rules
+  enum EnforcementType {
+    // This specifies that the rule is required and must be strictly enforced
+    Required = 0;
+    // This specifies that the rule is preferred and can be best effort
+    Preferred = 1;
+  }
+
+  // Specifies the rule enforcement policy. Can take values: required or preferred.
+  // (optional)
+  EnforcementType enforcement = 3;
+  // Defines the affinity rule for the volume placement
+  AffinityRule affinity = 4;
```

Comment:
the anti/affinity should be a part of the rule itself instead of having two of them in a rule ?

Reply:
I discussed this with vinod offline.

We have uses cases where we both both as part of the same rule. For e.g

So below, one of the replicas need to honor both the `affinity` and `antiAffinity` rule.

We can't add that as part of the 2nd rule since that will affect other replicas.

```
apiVersion: libopenstorage.org/v1alpha1
kind: VolumePlacementStrategy
metadata:
  name: data-collocate-with-log-placement
spec:
  rules:
    - affectedReplicas: 1 
      enforcement: required
      affinity:
        matchExpressions:
        - key: type
          operator: In
          values:
          - log
      antiAffinity:
        matchExpressions:
        - key: type
          operator: In
          values:
          - data
    - enforcement: preferred
      antiAffinity:
        matchExpressions:
        - key: media_type
          operator: NotIn
          values:
          - "SSD"
```

Reply:
it won't work for a third rule, because the rule has "1" affinity part and "1" anti-affinity part, because we now have rules within rules.